### PR TITLE
Add warning if multiple batch_sizes are found from ambiguous batch

### LINKED ISSUE #10093: `utilities.data.extract_batch_size` can silently fail -- maybe it needs to be less automatic?
## Motivation

Hey! So from what I understand `extract_batch_size` is used by `ResultCollection` to reduce/re-weight various logging metrics such as validation loss, etc.

We had an issue where our batch collation function aggregated Tensors into a list instead of stacking them (needed because of size mismatches), and `extract_batch_size` would recursively crawl through the batch object until it found a Tensor and returned its first dimention size -- which is wrong (the list length is correct batch size). So we would silently get "random" batch sizes and the val_loss logging becomes incorrect and then the LR scheduling / early stopping incorrectly triggers etc.

### Pitch

1. Maybe there should be some warning / note in the documentation (didn't find any)
2. If the batch object isn't a tensor (like a dict/dataclass), maybe try to `getattr(batch, 'batch_size')` and only then produce a warning and try to infer it automatically.

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10093

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -12,34 +12,57 @@
...
+    if isinstance(batch, (Iterable, dict)):
+        if isinstance(batch, dict):
+            batch = batch.values()
+
+        batch_sizes = list(flatten_list([_extract_batch_size(sample, recursive=True) for sample in batch]))
+        batch_size = batch_sizes[0]
+
+        if not all(bs == batch_size for bs in batch_sizes) and not recursive:
+            warning_cache.warn(
+                f"Lightning is trying to infer the `batch_size` from an ambiguous collection. The batch size we"
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:50]: haha... true!!

File: pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -12,34 +12,57 @@
...
+def _extract_batch_size(batch: BType, recursive=False) -> List[int]:
+    if isinstance(batch, torch.Tensor):
+        return [batch.size(0)]
+    if isinstance(batch, str):
+        return [len(batch)]
+    if isinstance(batch, (Iterable, dict)):
+        if isinstance(batch, dict):
+            batch = batch.values()
+
+        batch_sizes = list(flatten_list([_extract_batch_size(sample, recursive=True) for sample in batch]))
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:45]: hmm. yeah.. let me try some DSA :sweat_smile:

File: pytorch_lightning/utilities/data.py [hunk 3]
```diff
@@ -18,28 +18,55 @@
...
+
+
+def _extract_batch_size(batch: BType, batch_sizes) -> bool:
+    current_batch_size = None
+
+    if isinstance(batch, torch.Tensor):
+        current_batch_size = batch.size(0)
+    elif isinstance(batch, str):
+        current_batch_size = len(batch)
+    elif not isinstance(batch, (Iterable, dict)):
```
Comment (eladsegal) [inline pytorch_lightning/utilities/data.py:35]: @rohitgr7 
Instead of using `dict`, `Mapping` should be used.
Example for a case in which the batch will be `Mapping` but not `dict` is when using huggingface.
Currently, 9 is returned instead of 2 for this HF example code:
```
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
pt_batch = tokenizer(
    ["We are very happy to show you the 🤗 Transformers library.", "We hope you don't hate it."],
    padding=True,
    truncation=True,
    max_length=512,
    return_tensors="pt"
)
print(f"Inferred batch size is {extract_batch_size(pt_batch)}")
```

Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:35]: true.. it's a `UserDict`.. thanks you :)

File: CHANGELOG.md [hunk 4]
```diff
@@ -231,6 +231,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `ckpt_path` argument for `trainer.fit()` ([#10061](https://github.com/PyTorchLightning/pytorch-lightning/pull/10061))
 
 
+- Added warning if multiple batch_sizes are found from ambiguous batch ([#10247](https://github.com/PyTorchLightning/pytorch-lightning/pull/10247))
```
Comment (awaelchli) [inline CHANGELOG.md:234]:
```suggestion
- Added a warning if multiple batch sizes are found from ambiguous batch ([#10247](https://github.com/PyTorchLightning/pytorch-lightning/pull/10247))
```

File: pytorch_lightning/utilities/data.py [hunk 5]
```diff
@@ -18,28 +18,55 @@
...
 
 from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.warnings import WarningCache
 
 BType = Union[torch.Tensor, str, Mapping[Any, "BType"], Iterable["BType"]]
 
+warning_cache = WarningCache()
+
+
+def _extract_batch_size(batch: BType, batch_sizes) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/data.py:28]: isn't there a more elegant implementation using a Generator (yield)? 
Then there is no special return value needed, and the warning can be handled externally.

Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:28]: I'll try

**Review by tchaton — APPROVED** [review]
LGTM ! We can provide a more elegant version of it in another PR too.

File: pytorch_lightning/utilities/data.py [hunk 6]
```diff
@@ -64,9 +46,18 @@ def extract_batch_size(batch: BType) -> int:
...
         ``len(tensor)`` when found, or ``1`` when it hits an empty or non iterable.
     """
-    batch_sizes = []
-    _extract_batch_size(batch, batch_sizes)
-    return batch_sizes[0]
+    batch_size = None
+    for bs in _extract_batch_size(batch):
+        if batch_size is None:
+            batch_size = bs
+        elif batch_size != bs:
```
Comment (tchaton) [inline pytorch_lightning/utilities/data.py:53]: Neat !

**Review by tchaton — COMMENTED** [review]
Nice !

File: pytorch_lightning/utilities/data.py [hunk 7]
```diff
@@ -18,28 +18,46 @@
...
-        return extract_batch_size(sample)
-
-    return 1
+    batch_size = None
+    for bs in _extract_batch_size(batch):
+        if batch_size is None:
+            batch_size = bs
+        elif batch_size != bs:
+            warning_cache.warn(
+                f"Trying to infer the `batch_size` from an ambiguous collection. The batch size we"
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/data.py:55]:
```suggestion
                "Trying to infer the `batch_size` from an ambiguous collection. The batch size we"
```

File: pytorch_lightning/utilities/data.py [hunk 8]
```diff
@@ -18,28 +18,46 @@
...
 
 from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.warnings import WarningCache
 
 BType = Union[torch.Tensor, str, Mapping[Any, "BType"], Iterable["BType"]]
 
+warning_cache = WarningCache()
+
+
+def _extract_batch_size(batch: BType) -> int:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/data.py:28]:
```suggestion
def _extract_batch_size(batch: BType) -> Generator[int, None, None]:
```
did mypy not error?

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/utilities/data.py [hunk 9]
```diff
@@ -18,28 +18,46 @@
...
-        return len(batch)
-    if isinstance(batch, dict):
-        sample = next(iter(batch.values()), 1)
-        return extract_batch_size(sample)
-    if isinstance(batch, Iterable):
-        sample = next(iter(batch), 1)
-        return extract_batch_size(sample)
-
-    return 1
+    batch_size = None
```
Comment (awaelchli) [inline pytorch_lightning/utilities/data.py:49]:
```suggestion
    batch_size = 1
```
or annotation as `Optional[int]`

**Review by awaelchli — APPROVED** [review]

File: tests/utilities/test_data.py [hunk 10]
```diff
@@ -23,6 +23,22 @@ def test_extract_batch_size():
...
+    warning_cache.clear()
+
+    batch = {"test": [{"test": [torch.zeros(10, 10), torch.zeros(11, 10)]}]}
+    with pytest.warns(UserWarning, match="Trying to infer the `batch_size` .* we found is 10."):
+        extract_batch_size(batch)
+
+    warning_cache.clear()
+
+    batch = [{"test": torch.zeros(10, 10), "test_1": torch.zeros(11, 10)}]
+    with pytest.warns(UserWarning, match="Trying to infer the `batch_size` .* we found is 10."):
```
Comment (awaelchli) [inline tests/utilities/test_data.py:39]: I think this test needs a check that no warning is raised if the collection has a uniform batch size

File: pytorch_lightning/utilities/data.py (hunk 9 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:49]: it can never be optional ? since above it will return 1 in the last else condition. The reason I initalized it with None to first batch_size since there is a possibility of getting 1 in the first call itself.

File: pytorch_lightning/utilities/data.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:28]: why None here?

File: pytorch_lightning/utilities/data.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/data.py:49]: that's interesting, I didn't see that but that means mypy can infer that it's never None, i.e., that the for loop always runs?

Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:49]: I think yes, it will.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/data.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/data.py:49]: On my prs, it never wants to be smart...

Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:49]: lucky day maybe 😆
